File: src/Java/LeetcodeSolutions/src/main/java/leetcode/solutions/concrete/Solution_13_Roman_to_Integer.java
package leetcode.solutions.concrete;

import leetcode.solutions.*;
import org.jetbrains.annotations.NotNull;

import static leetcode.solutions.ProblemDifficulty.*;
import static leetcode.solutions.SolutionValidator.*;
import static leetcode.solutions.Complexity.*;

/**
 * @Problem: Given a roman numeral, convert it to an integer.
 * @Solution:
 * <ul>
 * <li>Iterate over the provided string, checking each character.
 * <li>Increment the result variable by the appropriate amount, but with a check for the following situations:
 * <ol>
 *  <li>I can be placed before V (5) and X (10) to make 4 and 9.
 *  <li>X can be placed before L (50) and C (100) to make 40 and 90.
 *  <li>C can be placed before D (500) and M (1000) to make 400 and 900.
 *  </ol>
 * </ul>
 * @Complexity: O(N)
 * @Author: Daniil Kuprianov
 */

public class Solution_13_Roman_to_Integer extends LeetcodeSolution {

    public Solution_13_Roman_to_Integer() { super(13, "Roman to Integer", EASY); }

    @ProblemSolution(complexity = O_N)
    private int romanToInt(@NotNull String s) {
        if (s.length() > 15 || (!s.contains("I") && !s.contains("V") && !s.contains("X") &&
                                !s.contains("L") && !s.contains("C") && !s.contains("D") &&
                                !s.contains("M"))
           )
            return -1;

        int result = 0;
        char charBefore = '0';
        for (var i = 0; i < s.length(); ++i) {
            if (i != 0) charBefore = s.charAt(i - 1);
            switch (s.charAt(i)) {
                case 'I':
                    ++result;
                    break;
                case 'V':
                    if (charBefore == 'I') {
                        result += 3;
                    } else {
                        result += 5;
                    }
                    break;
                case 'X':
                    if (charBefore == 'I') {
                        result += 8;
                    } else {
                        result += 10;
                    }
                    break;
                case 'L':
                    if (charBefore == 'X') {
                        result += 30;
                    } else {
                        result += 50;
                    }
                    break;
                case 'C':
                    if (charBefore == 'X') {
                        result += 80;
                    } else {
                        result += 100;
                    }
                    break;
                case 'D':
                    if (charBefore == 'C') {
                        result += 300;
                    } else {
                        result += 500;
                    }
                    break;
                case 'M':
                    if (charBefore == 'C') {
                        result += 800;
                    } else {
                        result += 1000;
                    }
            }
        }
        return result;
    }

    @Override
    @ProblemInputData
    public void run() {
        ASSERT_EQ(3,    romanToInt("III"));
        ASSERT_EQ(58,   romanToInt("LVIII"));
        ASSERT_EQ(1994, romanToInt("MCMXCIV"));
    }
}

File: src/Java/LeetcodeSolutions/src/main/java/leetcode/solutions/concrete/kotlin/Solution_13_Roman_to_Integer.kt
package leetcode.solutions.concrete.kotlin

import leetcode.solutions.*
import leetcode.solutions.ProblemDifficulty.*
import leetcode.solutions.SolutionValidator.*
import leetcode.solutions.Complexity.*

/**
 * __Problem:__ Given a roman numeral, convert it to an integer.
 *
 * __Constraints:__
 * - 1 <= num <= 3999

 * __Solution 1:__
 * - Iterate over the provided string, checking each character.
 * - Increment the result variable by the appropriate amount, but with a check for the following situations:
 *
 * 1) I can be placed before V (5) and X (10) to make 4 and 9.
 * 2) X can be placed before L (50) and C (100) to make 40 and 90.
 * 3) C can be placed before D (500) and M (1000) to make 400 and 900.
 *
 * __Complexity:__ O(N)
 * @see Solution_12_Integer_to_Roman
 * @author Daniil Kuprianov
 */

class Solution_13_Roman_to_Integer : LeetcodeSolution(EASY) {
    init { resolveConcreteSolutionInfo(this) }

    @ProblemSolution(complexity = O_N)
    private fun romanToInt(s: String): Int {
        var result = 0
        var charBefore = '0'
        for (i in s.indices) {
            if (i != 0) charBefore = s[i - 1]
            when (s[i]) {
                'I' -> {
                    ++result
                }
                'V' -> {
                    result += if (charBefore == 'I') 3 else 5
                }
                'X' -> {
                    result += if (charBefore == 'I') 8 else 10
                }
                'L' -> {
                    result += if (charBefore == 'X') 30 else 50
                }
                'C' -> {
                    result += if (charBefore == 'X') 80 else 100
                }
                'D' -> {
                    result += if (charBefore == 'C') 300 else 500
                }
                'M' -> {
                    result += if (charBefore == 'C') 800 else 1000
                }
            }
        }
        return result
    }

    @ProblemInputData
    override fun run() {
        ASSERT_EQ(3,    romanToInt("III"))
        ASSERT_EQ(58,   romanToInt("LVIII"))
        ASSERT_EQ(1994, romanToInt("MCMXCIV"))
    }
}
